[Docs] Terrain Developer Guide (#2122)
o3de/o3de.org: content/docs/user-guide/visualization/environments/terrain/terrain-developer-guide/architecture.md

```goat
                  .--------------.
                 |    Terrain     |
                  '--+---+-+---+-'
                     |   | |   |
                     |   | |    '------------------.
            .-------'    |  '--------.              |
           |             |            |             |
           v             |            v             |
  .--------+----------.  |  .---------+------.      | 
 |   Gradient Signal   | | |   Surface Data   |     |
  '------------+------'  |  '------+---------'      |
               |         |         |                |
               v         v         v                v
        .------+---------+---------+------.     .--------.
       |   LmbrCentral (Shape Components)  |   |   Atom   |
        '---------------------------------'     '--------'
```

The **Gradient Signal Gem** provides components that map data in the 0-1 range to world positions. The gradient components are used for defining height data and surface weight data. In particular, the **Image Gradient** provides a convenient workflow for importing and using heightmaps and "splat maps" (surface weight maps) that are generated from external terrain authoring tools.

> The **Surface Data Gem** provides a way to define conceptual geometric shells with sets of arbitrary tags which can then be referenced from other systems (like Dynamic Vegetation) that operate on generic surfaces. The terrain system uses the Surface Data tags as the naming system for terrain surface types. It also implements a Surface Data component to make the terrain show up as a surface within the Surface Data system. This enables the Dynamic Vegetation system to treat terrain as a "plantable surface".

Author: Would "conceptual geometric volumes" make more sense?

2. EBus encourages decoupling. We want each piece of the system to be able to operate independently and only interact through the API contracts defined on the EBus.
3. EBus enables functional replacements. As long as a component or system meets the EBus contract, it's possible to replace the functionality in discrete chunks.

Queries to the system should be able to run in parallel and only block on data modifications, and data modifications should run serially and block on both data queries and data modifications. The locking pattern is optimized around the assumption that terrain data is queried in a read-only fashion most of the time, and the data is loaded, unloaded, or modified just a small fraction of the time.

## Single system, multiple components

Another key aspect to the terrain system design is that it provides a singleton system as a single entry point for all terrain queries and management of global terrain settings. However, all of the terrain data and components that feed the system are split apart into discrete pieces. There are many reasons for both of these design decisions.

> Single terrain system:

Author: It's per level. The system gets activated when you add the Terrain World level component to a level. Each level in a project that doesn't contain that component won't have an active terrain system.

  * Data usage can be optimized by only placing data where it's needed.
  * Regions can be dynamically loaded and unloaded individually, making it easy to implement coarse-grained terrain streaming.
* Each region can have different subsets of features.
  * Regions where gameplay occurs can spawn terrain physics representations, but distant terrain regions that are for visuals only don't need to have physics enabled.
  * Aspects of terrain rendering can be disabled for server-side or machine learning simulations that don't require a visual representation of terrain.
* Different components can have different input data resolutions. The terrain system defines a single _output_ terrain grid and resolution, but the input data resolutions can be varied to optimize the data size and complexity based on the need for that region. Even within a region, two different data providers (such as height and surface type) can use two different resolutions based on what's appropriate for that type of data.
* The functionality can be replaced at a component-by-component level.

>   * The PhysX Heightfield component could be replaced with one that connects to another physics system such as Jolt or Bullet.

Author: Those links are correct, but I don't see any obvious trademark information. Would it be correct to refer to them as "Jolt Physics" and "Bullet Physics SDK", since that's how they are referred to on their main github pages?

>   * The PhysX Heightfield component could be replaced with one that connects to another physics system such as Jolt or Bullet.

Reviewer: Yes! Using the name they use to refer to themselves is a safe bet.

  * New data providers can be created for streaming satellite data, performing complex procedural generation, etc, and swapped in without needing to replace the rest of the terrain system.

In short, the singleton terrain system provides a single simplified API for runtime systems that need terrain data. The separated components provide better workflows and flexibility for content authors with multiple layers of data tuning and optimization.

## Component design

The specific functional and data separations in the terrain component design have been chosen to make it easy to control what functionality is active and to provide clear touchpoints for replacing functionality everywhere that it's likely to be useful.

### Level components

The terrain system uses three level components:

| Component | Description |
| - | - |
| **Terrain World** | Controls the base terrain system. The entire system is enabled or disabled by adding or removing this component. The configuration parameters that affect the entire terrain world exist on this component as well.|

> | **Terrain World Renderer** | Controls the terrain rendering. It requires the **Terrain World** component to exist, and it provides all of the rendering configuration that needs to be consistent across the entire terrain. This component is kept separate from **Terrain World** so that it's possible to define a conceptual terrain without requiring the overhead of rendering for use cases where a visual representation simply isn't necessary. Also, this provides a "replacement touchpoint" for creating different terrain renderers, such as a voxel-based implementation or a raytracing-based implementation.|

Reviewer: - For "...that needs to be...", the way I read this felt like there's emphasis on how the rendering _needs_ to be consistent, which then makes me wonder what that means and why it needs to be. Instead, I believe this sentence intends to indicate simply that the rendering configuration applies consistently throughout the entire terrain. If my assumption is right, see my suggestion below. 
- For the sentence "This component is kept separate...", broke up the sentence because it was getting too long to read easily.
```suggestion
| **Terrain World Renderer** | Controls the terrain rendering. It requires the **Terrain World** component to exist, and it provides all of the rendering configuration, which applies consistently across the entire terrain. This component is separated from **Terrain World** so that it's possible to define a conceptual terrain without requiring the overhead of rendering. This is useful for cases where a visual representation simply isn't necessary. Also, this provides a "replacement touchpoint" for creating different terrain renderers, such as a voxel-based implementation or a raytracing-based implementation.|
```

Author: I'll change the first one to "it provides all of the rendering configuration settings that apply globally to the entire terrain." The other part looks good to me.

> | **Terrain World Debugger** | Provides helpful debugging features for visualizing aspects of the terrain system. This component is kept separate from **Terrain World** so that it can easily be completely removed from shipping products.|

Reviewer: ```suggestion
| **Terrain World Debugger** | Provides helpful debugging features for visualizing aspects of the terrain system. This component is kept separated from **Terrain World** so that it's easy to completely remove it from shipping products.|
```

Author: Should it be "is separated" vs "is kept separated"?

Reviewer: Ah yes, I think I meant to remove "kept" as I did in other comments. Thanks for catching!

### Entity components

There are two types of terrain entity components - the "core" components that are used to define a terrain region and "auxiliary" components that operate on an area that can contain terrain regions.

Core Components:

| Component | Description |
| - | - |
| **Axis-Aligned Box Shape** | Defines the shape of a terrain region, including the min/max heights. |
| **Terrain Layer Spawner** | Declares that a terrain region exists in the shape bounds, and it has a specified priority. |
| **Terrain Height Gradient List** | (Optional) Defines the height data for the terrain region. |
| **Terrain Surface Gradient List** | (Optional) Defines the surface type and weight data for the terrain region. |

Auxiliary Components:

The "auxiliary" components only require a Box Shape on their entity, but not a Terrain Layer Spawner, so they can span across multiple terrain regions instead of being tied to a single one. They _can_ still appear on an entity with a Terrain Layer Spawner, it just isn't required.

| Component | Description |
| - | - |

> | **Terrain Macro Material** | Defines terrain base coloring for all terrain regions that fall within its volume. Macro materials aren't directly attached to single terrain regions so that they can be authored and streamed at different world resolutions than the terrain regions. |

Reviewer: ```suggestion
| **Terrain Macro Material** | Defines terrain base color for all terrain regions that fall within its volume. Macro materials aren't directly attached to single terrain regions so that they can be authored and streamed at different world resolutions than the terrain regions. |
```

Author: I went with "coloring" instead of "color" because "color" sounds a little too singular, and "colors" sounds more like a list of colors than an applied texture. I'm open to more suggestions here though. Maybe "Applies a terrain base color texture to all terrain regions that fall within its volume."?

Here are some examples of what the macro material does to the terrain:
![level-with-macro-color](https://user-images.githubusercontent.com/82224783/204674966-c935a3b1-d4a5-439c-87df-19f5ae8821b7.png)
![level-with-macro-color-up-close](https://user-images.githubusercontent.com/82224783/204675002-ed9e1ff5-1a06-4517-8808-24e36cae8585.png)

Reviewer: Yes I think "a terrain base color texture" is perfect. Texture is a clue to the reader what kind of asset is used.
